### DISCUSSION TITLE
创建自定义控件的方法

### DISCUSSION THREAD

**Discussion by AIKA1024** [BYJRK/DotNet-Discussions#316] [WPF 开发]
### 第一种：直接创建用户控件
第一种就是直接创建一个用户控件，这种xaml和cs代码结合的很紧密，cs代码订阅事件也十分方便，但是无法用TemplateBinding来绑定外面设置的属性，我看到一些项目里这种一般都叫Widgets了。
个人感觉不会有太多定制化的功能就会使用这种办法，像是专门设计来展示特定数据
![QQ_1722594143402](https://github.com/user-attachments/assets/2157f881-eaea-4268-a113-4ae7b09d23a2)
### 第二种：创建一个类继承某个控件，但是样式写在另外的资源字典里
![QQ_1722594529548](https://github.com/user-attachments/assets/509ab623-7c95-4ef2-ac8e-f0f9e939bcc0)
![QQ_1722594594259](https://github.com/user-attachments/assets/718f56ef-375f-4810-a125-93be15c83f95)
![QQ_1722594617233](https://github.com/user-attachments/assets/77b8b2d4-5735-4896-b0a8-43668382e179)

这种办法就可以使用TemplateBinding，但是给某个元素订阅事件特别麻烦，虽然可以在OnApplyTemplate事件上手动订阅事件。这种办法也无法使用设计器，只能看着xaml一顿打。
个人感觉像是真正的在做一个可以定制控件

**以上全部是我的个人猜想，毕竟我也没做过wpf的工作，都是自己自学的，在做控件的时候总是犹豫到底用哪种，总是觉得这个控件其实可以做的通用一点。**

或者说有其他十全十美的做法吗

**Comment by BYJRK** [comment]
`CustomControl`，它没有 XAML，只有 `.cs`，然后用 `Template` 的方式实现样式（自然可以使用 `TemplateBinding` 了）。它的样式最终会被合并到 `Themes/Generic.xaml` 中，从而可以在别的项目中被引用。

你甚至可以在 VS 中创建一个 CustomControl，然后通过文件模板中自动生成的长篇注释来大概了解它的使用方式。

https://www.wpftutorial.net/HowToCreateACustomControl.html

**Comment by KatzReveil** [comment]
用户控件仅仅适用于简单控件，为了实现代码复用。要实现复杂功能是只能用自定义控件的，微软写wpf和winui都是用自定义控件写的。

**Reply by BYJRK** [reply]
你说的不对。首先，你不能用“微软写wpf和winui都是用自定义控件写的”来论证“要实现复杂功能是只能用自定义控件的”。其次，单从复杂功能的角度来说，用户控件也丝毫不弱于自定义控件，甚至在提供了XAML之后反而能够比自定义控件更加灵活。这两者的主要区别在于用途。如果你想实现的是简单的对于多个已有控件的集合，或者是大到整个导航页面，那么应该使用用户控件；如果你想开发可以用于分享的控件库，那么你应该选择使用自定义控件，从而去控制对于Style、Template的相关机制。

**Reply by ZhuYajunFly** [reply]
> 用户控件仅仅适用于简单控件，为了实现代码复用。要实现复杂功能是只能用自定义控件的，微软写wpf和winui都是用自定义控件写的。

自定义控件主要专注于自身，用户控件倾向于整合一个对象的功能（比如一个阀门、电机），个人理解哈
